olm/operator-registry.md

```diff
@@ -0,0 +1,318 @@
+
+The final implementation goal of this enhancement involves this set of steps:
+
+1. Use a reference to a container image (also referred to here as the operator index) rather than app-registry namespaces to expose installable operators on a cluster
+2. Build an operator-registry database that serves the data required by OLM to drive install and UI workflows
+3. Have a way to optimize the build workflow of operator-registry databases (which currently drive OLM's workflows) from previous versions of the database plus new content (ex. new operator at new version) so that the database need not be built from scratch every time they are created.
+
+To start, let's define what the operator-registry does today. It:
+
+1. Takes a set of manifest files and outputs a database file.
+2. Takes a database file and serves a grpc api that serves content from that db
+3. Takes a configmap reference, builds a db and serves a grpc api that serves content from that db
+4. Takes an app-registry path/namespace reference, builds a db and serves a grpc api that serves content from that db
+
+Below are a set of implementation steps that will add new features to accomplish our goals:
+
+*Update the Operator Registry to insert incremental updates*
+
+Add create db, delete, and batch insert APIs to the model layer of operator-registry.
+
+Add a new set of operator registry commands to utilize those new APIs:
+
+`operator-registry create`
+    - inputs: none
+    - outputs: empty operator registry database
+    
+`operator-registry add`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add quay.io/community-operators/foo:sha256@abcd123 example.db`
+    
+`operator-registry add --batch`
+    - inputs: $operatorBundleImagePath, $dbFile
+    - outputs: updated database file
+    ex: `operator-registry add "quay.io/community-operators/foo:sha256@abcd123,quay.io/community-operators/bar:sha256@defg456" example.db`
+    
+`operator-registry delete`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without $operatorName included
+    ex: `operator-registry delete bar example.db`
+
+`operator-registry delete-latest`
+    - inputs: $operatorName, $dbFile
+    - outputs: updated database file without latest version of $operatorName included
+    ex: `operator-registry delete-latest foo example.db`
+    
+*Reference non latest versions of bundles by image digests*
+
+Currently the operator-registry database has a table to the effect of:
+
+OperatorBundle
+    -name: foo
+    -bundle: "{$jsonblob}"
+    
+We will add a field to this table that will reference the bundle image path and, when the operator is not the latest version of the default channel, will not include the bundle blob itself:
```

Comment:
Would we maybe just want to have both for the default/lastest version? could it hurt or what would be the reason to not have it I guess? 

It might be nice that it is available in the future, right?

Reply:
We will explicitly have both. This is just saying that for non latest we won't retain the blob. In either case we will need to have the image path, otherwise we will have no way of finding it later.